docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,244 @@
+    connection will switch to it) and **removes the "losing flag" from the in-memory store**.
+  - **The Function also proactively informs LNS1** that is not anymore the winning LNS for this
+    device. The reason why we do this proactively is that otherwise its Edge Hub will try to
+    reconnect to IoT Hub even if there is no more messages picked up from LNS1 (out of range roaming
+    client). Additionally, **LNS1 would not need to refresh its LoRaDeviceCache anymore for this
+    device**.
+    - Direct method and should be retried (e.g. in memory retries or durable Function): direct
+      method could fail due to the module being or Edge Hub being down, other connectivity issue
+      etc. Is in memory retries sufficient here? ❔
+  - If LNS1 in the meantime gets message B and contacts the FunctionBundler, it will let it know
+    that it lost the race for this frame counter and must therefore drop the message, **mark
+    itself as the losing LNS and close the connection**.
+
+#### Handling of Join requests
+
+Join requests in isolation currently do not have the connection stealing issue, as they rely on the Function for
+the DevNonce check. If the current LNS is not the preferred gateway, it drops the message
+immediately.
+
+We could also consider storing the information that we were the losing LNS locally on the LNS. The
+advantages of that are:
+
+- subsequent Data (or Join) requests do not potentially switch the connection
+- same handling as in the Data flow (therefore easier to follow)
+  
+Not sure the following section is relevant as both data and join flow need DevEui.
+The problem is that the Data message flow relies on the DevAddr while here we only have the DevEui.
+Ways to solve this:
+
+- Split the creation of the LoRaDevice from fetching the twin, create a dummy LoRaDevice and
+  mark our LNS as the losing one.
+- Store outside of LoRaDevice a data-structure with DevEui: flag (winning or losing) and inject
+  also on the data message handler
+- Do nothing and accept that a one-off connection stealing on the first data message can happen.
+
+#### Delay on the LNS itself or on the Function ❔
+
+What are the scenarios that it's better that the function implements the delay?
+
+Disadvantages of using a Task.Delay on the Function:
+
+- Observability: potentially we are messing up the measurements of the Function duration.
+- Keeps the HTTP connection between the LNS-Function open for more time.
+  
+#### Device reset ❔
+
+What happens if we have a device reset between message A and B? We save the twin
+immediately (irrelevant but how does the Function get it for the deduplication decision?)
+
+Unclear what is the relationship between the twin and the cache on the function: in the cache we
+keep the framecounter down from the twin but not the fcnt up ❔
+
+- If both LNSs receive message B, LNS1 SHOULD reset the frame counter. LNS2 then
```

Comment:
I think this was a problem before. Both would perform the reset. Not sure this is in scope here?

Reply:
Resolving this as the section was re-written and the comment isn't applicable.